Add key origin support to descriptors

sipa opened the pull request:
This adds support for [key origin](https://gist.github.com/sipa/e3d23d498c430bb601c5bca83523fa82#key-origin-identification) information to the descriptor parser, and exposes the resulting key path information through `FlatSigningProvider`.

There is no observable functionality from this right now, except having the `scantxoutset` RPC accept descriptors that include key origin information.

Longer term this feature helps with a potential descriptors-based walletless PSBT updater, or for importing hardware wallet xpubs (once the wallet can import descriptors).

sipa committed:
Add key origin support to descriptors

sipa committed:
Add tests for key origin support

sipa commented on src/script/descriptor.cpp:
```diff
@@ -41,7 +41,7 @@ struct PubkeyProvider
 
     /** Derive a public key. */
-    virtual bool GetPubKey(int pos, const SigningProvider& arg, CPubKey& out) const = 0;
+    virtual bool GetPubKey(int pos, const SigningProvider& arg, std::pair<CPubKey, KeyOriginInfo>& out) const = 0;
```
Agree, that's actually simpler. Done.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -272,15 +312,19 @@ class MultisigDescriptor : public Descriptor
-            out.pubkeys.emplace(key.GetID(), std::move(key));
+        std::vector<CPubKey> pubkeys;
+        pubkeys.reserve(entries.size());
+        for (auto& entry : entries) {
```
I don't like interfaces that populate output argument when failure occurs.

promag replied:
Ah agree on that.

Sjors commented on src/test/descriptor_tests.cpp:
```diff
@@ -100,9 +103,13 @@ void Check(const std::string& prv, const std::string& pub, int flags, const std:
                     BOOST_CHECK_MESSAGE(SignSignature(Merge(keys_priv, script_provider), spks[n], spend, 0, 1, SIGHASH_ALL), prv);
                 }
             }
```
Maybe add a comment:
`// Check if origin information (master key fingerprint, derivation path) is present`

It's actually not obvious to me how this test checks that an origin string passed into the test is actually stored correctly.

sipa replied:
Added some comments.

Sjors commented on src/test/descriptor_tests.cpp:
```diff
@@ -135,20 +142,22 @@ BOOST_AUTO_TEST_CASE(descriptor_test)
-    Check("wpkh(xprv9vHkqa6EV4sPZHYqZznhT2NPtPCjKuDKGY38FBWLvgaDx45zo9WQRUT3dKYnjwih2yJD9mkrocEZXo1ex8G81dwSM1fwqWpWkeS3v86pgKt/1/2/*)", "wpkh(xpub69H7F5d8KSRgmmdJg2KhpAK8SR3DjMwAdkxj3ZuxV27CprR9LgpeyGmXUbC6wb7ERfvrnKZjXoUmmDznezpbZb7ap6r1D3tgFxHmwMkQTPH/1/2/*)", RANGE, {{"0014326b2249e3a25d5dc60935f044ee835d090ba859"},{"0014af0bd98abc2f2cae66e36896a39ffe2d32984fb7"},{"00141fa798efd1cbf95cebf912c ...
-    Check("sh(wpkh(xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi/10/20/30/40/*'))", "sh(wpkh(xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8/10/20/30/40/*'))", RANGE | HARDENED, {{"a9149a4d9901d6af519b2a23d4a2f51650fcba87ce7b87"},{"a914bed59fc0024fae941d6e20a3b44a109ae ...
-    Check("combo(xprvA2JDeKCSNNZky6uBCviVfJSKyQ1mDYahRjijr5idH2WwLsEd4Hsb2Tyh8RfQMuPh7f7RtyzTtdrbdqqsunu5Mm3wDvUAKRHSC34sJ7in334/*)", "combo(xpub6FHa3pjLCk84BayeJxFW2SP4XRrFd1JYnxeLeU8EqN3vDfZmbqBqaGJAyiLjTAwm6ZLRQUMv1ZACTj37sR62cfN7fe5JnJ7dh8zL4fiyLHV/*)", RANGE, {{"2102df12b7035bdac8e3bab862a3a83d06ea6b17b6753d52edecba9be46f5d09e076ac","76a914f90e3178ca25f2c808dc76624032d352fdbdfaf288ac","0014f ...
+    Check("combo(01234567:xprvA1RpRA33e1JQ7ifknakTFpgNXPmW2YvmhqLQYMmrj4xJXXWYpDPS3xz7iAxn8L39njGVyuoseXzU6rcxFLJ8HFsTjSyQbLYnMpCqE2VbFWc)", "combo(01234567:xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHWkY4ALHY2grBGRjaDMzQLcgJvLJuZZvRcEL)", SIGNABLE, {{"2102d2b36900396c9282fa14628566582f206a5dd0bcc8d5e892611806cafb0301f0ac","76a91431a507b815593dfc51ffc7245ae7e5aee30 ...
```
In BIP32 this test vector is `m/0/2147483647'/1/2147483646'` derived from master seed `fffcf9f...`. It would be better to use that for the origin info.

sipa replied:
Meh.

This is not testing BIP32; it's just reusing the BIP32 tests because that was an easy place to find example xpubs/xprvs.

Sjors reviewed (commented):
Concept ACK.

For me this would be easier to review when combined with a practical usage, like creating a PSBT with the origin info still in it. That could just be a WIP PR as it might involve quite a few changes, since descriptors currently are only used with `scantxoutset`.

practicalswift commented on src/test/descriptor_tests.cpp:
```diff
@@ -43,9 +43,12 @@ std::string MaybeUseHInsteadOfApostrophy(std::string ret)
 }
 
-void Check(const std::string& prv, const std::string& pub, int flags, const std::vector<std::vector<std::string>>& scripts)
+static const std::set<std::vector<uint32_t>> ONLY_EMPTY{{}};
```
```
warning: 'ONLY_EMPTY' is a static definition in anonymous namespace; static is redundant here [readability-static-definition-in-anonymous-namespace]
```

sipa replied:
Fixed.

practicalswift commented on src/script/sign.h:
```diff
@@ -58,10 +58,12 @@ struct FlatSigningProvider final : public SigningProvider
 
     bool GetCScript(const CScriptID& scriptid, CScript& script) const override;
     bool GetPubKey(const CKeyID& keyid, CPubKey& pubkey) const override;
+    bool GetKeyOrigin(const CKeyID& id, KeyOriginInfo& info) const override;
```
```
warning: function 'FlatSigningProvider::GetKeyOrigin' has a definition with different parameter names [readability-inconsistent-declaration-parameter-name]
```

sipa replied:
Fixed.

practicalswift commented on src/script/descriptor.cpp:
```diff
@@ -221,8 +260,10 @@ class SingleKeyDescriptor final : public Descriptor
+        if (!m_provider->GetPubKey(pos, arg, key, info)) return false;
         output_scripts = std::vector<CScript>{m_script_fn(key)};
+        out.origins.emplace(key.GetID(), std::move(info));
         out.pubkeys.emplace(key.GetID(), std::move(key));
```
```
warning: std::move of the variable 'key' of the trivially-copyable type 'CPubKey' has no effect; remove std::move() [hicpp-move-const-arg]
```

sipa replied:
Fixed.

practicalswift commented on src/script/descriptor.cpp:
```diff
@@ -272,15 +313,19 @@ class MultisigDescriptor : public Descriptor
+        for (auto& entry : entries) {
+            pubkeys.push_back(entry.first);
+            out.origins.emplace(entry.first.GetID(), std::move(entry.second));
+            out.pubkeys.emplace(entry.first.GetID(), std::move(entry.first));
```
```
warning: std::move of the expression of the trivially-copyable type 'CPubKey' has no effect; remove std::move() [performance-move-const-arg]
```

sipa replied:
Fixed.

Sjors commented:
Perhaps @achow101 can use this in #14021?

achow101 commented:
@Sjors I don't think this would be useful there. Instead we should have some new command that lets us import descriptors into the wallet.

practicalswift commented on src/script/descriptor.cpp:
```diff
@@ -139,6 +173,11 @@ class BIP32PubkeyProvider final : public PubkeyProvider
+        CKeyID keyid = m_extkey.pubkey.GetID();
+        std::copy(keyid.begin(), keyid.begin() + 4, info.fingerprint);
+        info.path = m_path;
+        if (m_derive == DeriveType::UNHARDENED) info.path.push_back(pos);
```
```
2018-09-25 20:23:55 clang(pr=14150): script/descriptor.cpp:179:69: warning: implicit conversion changes signedness: 'int' to 'std::vector<unsigned int, std::allocator<unsigned int> >::value_type' (aka 'unsigned int') [-Wsign-conversion]
```

sipa replied:
Fixed.

sipa committed:
Update documentation to incude origin information

sipa commented:
Rebased, and expanded documentation to include key origin information.

ryanofsky commented:
Was there a problem with the last rebase? It seems like some changes that were previously made, like https://github.com/bitcoin/bitcoin/pull/14150#discussion_r215631128, have been reverted.

MeshCollider commented:
Concept ACK

sipa commented:
@ryanofsky Wow, indeed, not sure how that happened.

I went through all comments again, and reapplied the fixed I made earlier where necessary. I've also expanded the documentation change a bit.

ryanofsky commented on doc/descriptors.md:
```diff
@@ -34,6 +34,7 @@ Output descriptors currently support:
 - `sh(wsh(multi(1,03f28773c2d975288bc7d1d205c3748651b075fbc6610e58cddeeddf8f19405aa8,03499fdf9e895e719cfd64e67f07d38e3226aa7b63678949e6e49b241a60e823e4,02d7924d4f7d43ea965a465ae3095ff41131e5946f3c85f79e44adbcf8e27e080e)))` represents a P2SH-P2WSH *1-of-3* multisig.
 - `pk(xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8)` refers to a single P2PK output, using the public key part from the specified xpub.
 - `pkh(xpub68Gmy5EdvgibQVfPdqkBBCHxA5htiqg55crXYuXoQRKfDBFA1WEjWgP6LHhwBZeNK1VTsfTFUHCdrfp1bgwQ9xv5ski8PX9rL2dZXvgGDnw/1'/2)` refers to a single P2PKH output, using child key *1'/2* of the specified xpub.
+- `pkh(d34db33f/44'/0'/0':xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHWkY4ALHY2grBGRjaDMzQLcgJvLJuZZvRcEL/1/*)` refers to a chain of P2PKH outputs, but additionally specifies that the specified xpub is a child of a master with fingerprint `d34db33f`, and derived using path `44'/0'/0'`.
```
This syntax is slightly weird to me because if you're looking at an expression like:

`fingerprint/path1:xpub/path2`

The `fingerprint:path1` part only describes `xpub`, not `xpub/path2`, but there's no logical precedence of `:` and `/` operators that would lead me to see:

(`fingerprint/path1:xpub`)`/path2`

instead of:

(`fingerprint/path1`):(`xpub/path2`)

Maybe this is an unimportant aesthetic objection, but I'd almost rather there be an explicit `origin` function like:

`origin(fingerprint/path1, xpub)/path2`

or 

`origin(xpub, fingerprint/path1)/path2`

that at least would group things together correctly.

ryanofsky commented on src/script/descriptor.cpp:
```diff
@@ -484,6 +532,25 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(const Span<const char>& sp, bool per
+    if (colon_split.size() > 2) return nullptr;
+    if (colon_split.size() == 1) return ParsePubkeyInner(colon_split[0], permit_uncompressed, out);
+    auto slash_split = Split(colon_split[0], '/');
+    if (slash_split[0].size() != 8) return nullptr;
```
Could add CheckUnparsable checks for this condition and the !IsHex one below. Also maybe the !provider one.

sipa replied:
Done.

ryanofsky reviewed (approved):
utACK 3fdc704f06bb5618cb63c248b20447026509e00d. Left some comments below that aren't critical and you should feel free to ignore.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -56,16 +56,50 @@ struct PubkeyProvider
+
+    std::string OriginString() const
+    {
+        return HexStr(m_origin.fingerprint, m_origin.fingerprint + 4) + FormatKeyPath(m_origin.path);
```
I've added `std::begin`/`std::end` in a number of places, and replaced the other 4s with `sizeof(info.fingerprint)`. In C++17 we can use `std::size(info.fingerprint)` which would remain correct if the elements of the fingerprint weren't `sizeof()==1`, but that's too much work now.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -272,15 +313,19 @@ class MultisigDescriptor : public Descriptor
-            pubkeys.push_back(key);
+            std::pair<CPubKey, KeyOriginInfo> keyinfo;
+            if (!p->GetPubKey(pos, arg, keyinfo.first, keyinfo.second)) return false;
+            entries.emplace_back(std::move(keyinfo));
```
Done.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -484,6 +532,25 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(const Span<const char>& sp, bool per
+    if (!IsHex(fpr_hex)) return nullptr;
+    auto fpr_bytes = ParseHex(fpr_hex);
+    KeyOriginInfo info;
+    std::copy(fpr_bytes.begin(), fpr_bytes.end(), info.fingerprint);
```
Done, both.

sipa commented:
@ryanofsky Addressed all your comments except the syntax issue, which we should probably discuss at the PR level rather than buried inside a code comment.

I agree that `fpr/path:key/path` doesn't perfectly map with intuition about precedence of operation. On the other hand, I'd prefer to keep "key operations" as pure syntax, separated from "script operations" which are functions.

After thinking a bit more about it, how do you feel about `fpr/path(key)/path`? That way the entire expression maintains the order of path elements, and it just looks like you're "clarifying" the actual key somewhere inside the path itself, rather than as a separate thing prefixed onto it.

ryanofsky reviewed (commented):
utACK 71cbcd9da8b31cb7cbe19001df32c8e1c46a603b. Only changes since last review were suggested changes (replacing +4's, adding a few test cases and a comment).

---

I think I like both of the `[fpr/path]key/path` and `fpr/path[key]/path` syntaxes more than the current `fpr/path:key/path` syntax and more than my `origin()` suggestion, and I agree that the  `[fpr/path]key/path` version is probably more understandable.

I might also propose a more generalized syntax for adding attributes to keys like:

`key[origin=fpr/path, device=myhardwarewallet]/path`

Which might have other uses like:

`key[birthday=timestamp]`

But this might be too heavyweight. Anyway, you've thought about this a lot more than I have and I think whatever syntax you like, including the current syntax, should be fine.

sipa commented:
@ryanofsky A PSBT updater needs to add key origin information to the PSBT file, or future signers may not be able to find the key to sign with.

I really think this information needs to be part of descriptors for them to be generally useful. Just listing the public keys is useless if those using the information don't know how to find the private key corresponding to said public key. In a way, that derivation information should from our point of view be treated as part of the public information about that key.

sipa commented:
Switched to the `[fpr/path]key/path` notation, updating the parser, serializer, tests, and documentation.

As far as other attributes go, I'm not sure descriptors are the right place (this may be getting philosophical, though):
* a timestamp is more a descriptor-level property than a key specific thing (you care about when any output generated by the descriptor can be used at the earliest, not differences in timestamps between the individual keys inside one).
* the name of a hardware device is not portable (if included in descriptors, you wouldn't be able to copy the descriptor to another software/system that knows the hw device by a different name)

If there are generally useful properties of keys or path elements later we can still adopt such a syntax, though; it looks pretty readable.

ryanofsky commented on doc/descriptors.md:
```diff
@@ -52,14 +53,20 @@ Descriptors consist of several types of expressions. The top level expression is
-  - Instead of a `'`, the suffix `h` can be used to denote hardened derivation.
+- Optionally, key origin information, consisting of:
+  - An open bracket `[`
+  - 8 hex characters for the fingerprint
```
Would it be clearer to replace "fingerprint" with "extended key identifier"? Could also link to https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#key-identifiers if so.

instagibbs replied:
It's specifically "The first 32 bits of the identifier are called the key fingerprint.", not the entire identifier.

I think "master fingerprint" is more specific if we're going to change this.

sipa replied:
It isn't required to be the master fingerprint; just whatever key the device with the private key is on needs to derivate from. I guess most commonly that will be the master, but it could be a child key as well.

sipa replied:
I've expanded the comment slightly anyway.

ryanofsky reviewed (approved):
utACK cfc812992da86d00406072fe8d3e7e2611c67882. Only change since last review is syntax update.

---

Thanks for the explanations. I can see why you wouldn't want to make birthdays a property of keys, and it's also very clear that a PBST API which takes descriptors will be a lot clumsier if key origin information has to be passed in a separate table, instead of just embedded in the descriptors.

instagibbs commented:
concept ACK, if rebase is fixed I'll review further

ryanofsky commented:
> if rebase is fixed I'll review further

I don't think there's a problem with current version of this PR (cfc812992da86d00406072fe8d3e7e2611c67882). The earlier comment https://github.com/bitcoin/bitcoin/pull/14150#issuecomment-430359968 was referring to 605367580b582a6e9510a06670cfd4898168cea5.

instagibbs commented:
I literally just checked yesterday, perhaps taking crazy pills.

instagibbs commented on doc/descriptors.md:
```diff
@@ -107,6 +114,33 @@ Whenever a public key is described using a hardened derivation step, the
+corresponds with the master key `m`. Unfortunately, since there are
+hardened derivation steps that follow the xpub, this descriptor does not
+let you compute scripts without access to the corresponding private keys.
+Instead, it should be written as `xpub.../1/*`, where xpub corresponds to
```
ultra-micro nit: you used `i` earlier in the paragraph, and now `*` to describe the chain.

sipa replied:
Fixed, this was from a much earlier version.

instagibbs commented on test/functional/rpc_scantxoutset.py:
```diff
@@ -82,7 +82,7 @@ def run_test(self):
         assert_equal(self.nodes[0].scantxoutset("start", [ "combo(tprv8ZgxMBicQKsPd7Uf69XL1XwhmjHopUGep8GuEiJDZmbQz6o58LninorQAfcKZWARbtRtfnLcJ5MQ2AtHcQJCCRUcMRvmDUjyEmNUWwx8UbK/1/1/1500)"])['total_amount'], Decimal("16.384"))
         assert_equal(self.nodes[0].scantxoutset("start", [ "combo(tpubD6NzVbkrYhZ4WaWSyoBvQwbpLkojyoTZPRsgXELWz3Popb3qkjcJyJUGLnL4qHHoQvao8ESaAstxYSnhyswJ76uZPStJRJCTKvosUCJZL5B/1/1/0)"])['total_amount'], Decimal("4.096"))
-        assert_equal(self.nodes[0].scantxoutset("start", [ "combo(tpubD6NzVbkrYhZ4WaWSyoBvQwbpLkojyoTZPRsgXELWz3Popb3qkjcJyJUGLnL4qHHoQvao8ESaAstxYSnhyswJ76uZPStJRJCTKvosUCJZL5B/1/1/1)"])['total_amount'], Decimal("8.192"))
+        assert_equal(self.nodes[0].scantxoutset("start", [ "combo([abcdef88/1/2'/3/4h]tpubD6NzVbkrYhZ4WaWSyoBvQwbpLkojyoTZPRsgXELWz3Popb3qkjcJyJUGLnL4qHHoQvao8ESaAstxYSnhyswJ76uZPStJRJCTKvosUCJZL5B/1/1/1)"])['total_amount'], Decimal("8.192"))
```
Could there be a comment to point out what it's doing? As a reader of the test this is basically lost in the pile.

sipa replied:
Not sure what you're asking. I've added 2 lines of comment on this block of tests.

Or is it specifically about this including of the key origin in the test? It has no effect, as for now key origins are not observable anyway. So this is just testing that you can in fact specify one, and nothing breaks.

instagibbs replied:
I meant to a reader of the entire test itself, not the diff.

instagibbs reviewed (approved):
utACK

sipa commented on src/script/descriptor.cpp:
```diff
@@ -484,6 +533,28 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(const Span<const char>& sp, bool per
+/** Parse a public key including origin information (if enabled). */
+std::unique_ptr<PubkeyProvider> ParsePubkey(const Span<const char>& sp, bool permit_uncompressed, FlatSigningProvider& out)
+{
+    auto colon_split = Split(sp, ']');
```
Fixed.

sipa commented on doc/descriptors.md:
```diff
@@ -52,14 +53,20 @@ Descriptors consist of several types of expressions. The top level expression is
+- Optionally, key origin information, consisting of:
+  - An open bracket `[`
+  - 8 hex characters for the fingerprint
+  - Followed by zero or more "/NUM" or "/NUM'" path elements to indicate unhardened or hardened derivation steps between the fingerprint and the key that follows
```
Fixed.

sipa commented:
Addressed further comments in separate commits, will squash when ready.

instagibbs commented:
re-ACK

MeshCollider commented on src/script/descriptor.cpp:
```diff
@@ -272,15 +313,20 @@ class MultisigDescriptor : public Descriptor
-            CPubKey key;
-            if (!p->GetPubKey(pos, arg, key)) return false;
-            pubkeys.push_back(key);
+            std::pair<CPubKey, KeyOriginInfo> keyinfo;
```
I think you meant to use this below

MeshCollider commented on doc/descriptors.md:
```diff
@@ -52,14 +53,20 @@ Descriptors consist of several types of expressions. The top level expression is
+- Optionally, key origin information, consisting of:
+  - An open bracket `[`
+  - 8 hex characters for the fingerprint
+  - Followed by zero or more "/NUM" or "/NUM'" path elements to indicate unhardened or hardened derivation steps between the fingerprint and the key that follows
```
Using backticks (`) rather than quotes around /NUM and /NUM' is better for clarity, like below

Also tiny nit, typo in commit message.

meshcollider reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/pull/14150/commits/f4a8ccb6f754632167113f60f9764d60788bc89c

ryanofsky commented:
utACK 91cff98264ba5c6a86ca5d18b2591f759cffe793, just documentation fixes and two variable cleanups

sipa commented:
Squashed fixups.

sipa commented:
Rebased afa9224b14 -> 8afb166875

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/commit/8afb166875ffa2c6fcc32068437588ab1e1d109c

ryanofsky reviewed (approved):
utACK 8afb166875ffa2c6fcc32068437588ab1e1d109c. Only change since last review is rebase/squash, resolving a merge conflict with #14161 in the documentation
